internal/suppliers/index_mh_iter.go

```diff
@@ -0,0 +1,50 @@
+package suppliers
+
+import (
+	"context"
+	"io"
+
+	provider "github.com/filecoin-project/indexer-reference-provider"
+	"github.com/ipld/go-car/v2/index"
+	"github.com/multiformats/go-multihash"
+)
+
+var _ provider.MultihashIterator = (*indexMhIterator)(nil)
+
+type indexMhIterator struct {
+	mhch  chan multihash.Multihash
+	errch chan error
+}
+
+func newIndexMhInterator(ctx context.Context, idx index.IterableIndex) *indexMhIterator {
+	errch := make(chan error, 1)
+	mhch := make(chan multihash.Multihash, 1)
+	go func() {
+		if err := idx.ForEach(func(mh multihash.Multihash, _ uint64) error {
+			select {
+			case mhch <- mh:
+				return nil
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}); err != nil {
+			errch <- err
+		}
+		close(mhch)
+		close(errch)
+	}()
+	return &indexMhIterator{
+		mhch:  mhch,
+		errch: errch,
+	}
+}
+
+func (i indexMhIterator) Next() (multihash.Multihash, error) {
+	if next, ok := <-i.mhch; ok {
+		return next, nil
+	}
+	if err, ok := <-i.errch; ok {
+		return nil, err
+	}
```

Comment:
I'd argue you need to check the error channel before the multihash channel. If we've encountered an error, such as a timeout, we want to report that immediately - instead of returning the last buffered multihash, and then the error. Probably document that in the code, too.